### PR TITLE
Feature/user permissions

### PR DESCRIPTION
**Pull request by jesusgonzalezrivera**
Permit user to execute operations over datasets, widgets and metadata depending on their role

## Testing instructions

If you already have a user with admin permissions on the API and FA, I recommend you to create another user with the same email like this: `your_user+sa@vizzuality.com`. You can change the permissions of this new user with your existing admin user, so you can see the new permissions when the user is a Publisher or a Site Admin.

### Datasets

**Site Admin**

1. On the list of datasets, you should be able to see all datasets with edit and remove buttons available
2. Access to the edition of a dataset which does not pertain to your current user and complete all steps
3. You should see the changes in the dataset
4. Remove the edited dataset and check that it does not appear in the list
5. Access to the edition of a dataset of the current user and complete all steps
6. You should see the changes in the dataset
7. Remove the edited dataset and check that it does not appear in the list

**Publisher**

1. On the list of datasets, you should be able to see all datasets. However, you can only see the edit button available on those datasets which pertains to the current user.
2. Access to the edition of a dataset and complete all steps
3. You should see the changes on the dataset
4. Remove the edited dataset and check that it does not appear in the list

### Widgets

**Site Admin**

1. On the list of widgets, you should be able to see all widgets with edit and remove buttons available
2. Access to the edition of a widget which does not pertain to your current user and complete all steps
3. You should see the changes in the widget
4. Remove the edited widget and check that it does not appear in the list
5. Access to the edition of a widget of the current user and complete all steps
6. You should see the changes in the widget
7. Remove the edited widget and check that it does not appear in the list

**Publisher**

1. On the list of widgets, you should be able to see all the widgets. However, you can only see the edit button available on those widgets which pertains to the current user.
2. Access to the edition of a widget and complete all steps
3. You should see the changes on the widget
4. Remove the edited widget and check that it does not appear in the list

## Pivotal Tracker

[https://www.pivotaltracker.com/story/show/170070837](https://www.pivotaltracker.com/story/show/170070837)

### CODE REVIEW

**Review by clementprdhomme — CHANGES_REQUESTED** [review]
Hi Jesús,

I found several issues, some of them directly related to the feature, but others we might fix separately.

Issues directly related to the feature:
- When I log in as a publisher, the quick links component shows links to the “Admin” and “Contexts”. If you click on them, a rails error is thrown. I think these links should be hidden for the publishers.
- ~When I log in as a publisher and I create a dataset, I can't edit or delete it~.
- ~When I log in as a publisher and I create a widget, I can't delete it.~

Issues unrelated to the feature:
- When I edit a dataset, the “Continue” button doesn't work ([task](https://www.pivotaltracker.com/story/show/171625941)).
- When I edit a dataset, make changes to the “Metadata” step, go to the next one by clicking the tab and save, the changes are not saved ([task](https://www.pivotaltracker.com/story/show/171626007)).
- When I delete dataset that hasn't any widget, the modal that warns me contains an empty list of widgets ([task](https://www.pivotaltracker.com/story/show/171626065)).

File: app/assets/javascripts/views/shared/quickLinksView.js [hunk 1]
```diff
@@ -37,6 +37,21 @@
...
+      let userSiteRole = gon.global.user.user_site_roles.filter((userSite) => {
+        return userSite.site_slug === siteSlug;
+      });
+      userSiteRole = userSiteRole.length ? userSiteRole[0].role : "";
+      if (userSiteRole === 'publisher') {
+        for(var i = this.options.links.length; i--;) {
+          if (this.options.links[i]['id'] === 'admin' ||
+              this.options.links[i]['id'] === 'contexts') {
+            this.options.links.splice(i, 1);
+          }
```
Comment (clementprdhomme) [inline app/assets/javascripts/views/shared/quickLinksView.js]: You could simplify the code by this:
```js
this.options.links = this.options.links.filter(link => link.id !== 'admin' && link.id !== 'contexts');
```

But feel free to keep as it is 😄.

**Review by clementprdhomme — CHANGES_REQUESTED** [review]
Hi Jesús,

I don't think your fix cover all the cases.

Currently, only on URLs like this one, the admin and contexts menus are hidden: `/management/sites/:siteSlug/site_pages`.

But when the user logs in, they land on this page: `/management`. And there, they can still see the links to the admin and contexts sections.

What's also disturbing is that if I go directly to `/admin`, I'm able to edit sites, despite being only a content manager. You should not only hide the links, but also prevent managers from being able to access those pages.

File: app/assets/javascripts/views/shared/quickLinksView.js (hunk 1 above)
Comment (jesusgonzalezrivera) [reply app/assets/javascripts/views/shared/quickLinksView.js]: Too much time without using javascript 💯

**Review by clementprdhomme — APPROVED** [review]

**Pull request merged**